Remove serialize='pickle' from reports and saved_reports

minhaminha opened the pull request:
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Jira Ticket
](https://dimagi-dev.atlassian.net/browse/SAAS-11385)
This PR is part of a series of PRs aimed at removing our use of pickling in celery tasks due to a lack of support that is blocking our ability to upgrade celery.

Both reports and saved_reports are in this one pr since these two apps work closely with one another (refactoring a reports task required refactoring a saved_reports task).

The tasks refactored were:

In reports:
`export_all_rows_task(ReportClass, report_state, recipient_list=None, subject=None)` where `ReportClass` is changed to the class' `slug` where it then finds the correct class in a dictionary of imported report classes in the task. `report_state` was largely serializable except the `datespan` object which was removed and reformed in the task using `startdate` and `enddate` as strings. 

`build_form_multimedia_zipfile(domain, export_id, es_filters, download_id, owner_id)` where es_filters is a list of ExportFilter objects that have the method `to_es_filter` that converts itself to JSON compatible form (as referenced [here](https://github.com/dimagi/commcare-hq/pull/30338)).

`build_form_multimedia_zip(domain, export_id, datespan, user_types, download_id, owner_id)` which is deprecated by the previous task and removed. 

In saved_reports:
`send_report(notification_id)` and  `send_report_throttled(notification_id)` didn't need any changes beyond removing serializer='pickle'.

`send_email_report(self, recipient_emails, domain, report_slug, report_type, request_data, once, cleaned_data)` where `recipient_emails` is a list of strings, `domain`, `report_slug`, `report_type` are strings, `once` is a bool and `cleaned_data` is a dictionary of strings. From `request_data`, its 'GET' value was a QueryDict which was converted to a regular dictionary of strings before passing into the task and remade as a QueryDict object in the task and it's `datespan` object removed and reformed in the task using `startdate` and `enddate` as strings, which could be converted to json and back with `iso_string_to_datetime`. 


## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### QA Plan

[QA Ticket
](https://dimagi-dev.atlassian.net/browse/QA-3639)
<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

## Final self-reflection
- [x] My safety story above is convincing and gives me confidence that this PR will not introduce a regression

minhaminha committed:
Apply to_es_filter() before passing to celery

minhaminha committed:
Rework export_all_rows_task to be called with strings

minhaminha committed:
Remove build_form_multimedia_zip

minhaminha committed:
Remove pickles from saved_reports

Remove pickles from 3 tasks in saved_reports + remove outdated un-pickling function

minhaminha commented:
> Did a first pass, but need to look at this some more. I can confidently say this will require QA as this is an area I've messed up in the past 🙃. Specifically sending scheduled reports.

Will do! I was going to wait until I'm able to get all tests to pass but I think I've gotten stuck. Do you have any idea why it might raise `corehq.elastic.ESError: TransportError(404, 'index_not_found_exception', 'no such index')` for one of the tests when the index exists? It seems to only happen when trying to retrieve data NOT in a celery task (calling the task with .delay() vs without it). It feels like I'm missing something.

minhaminha committed:
Test fix

minhaminha committed:
Test fix 2

proteusvacuum commented:
Hey @minhaminha , do you think these sentry errors might be related to this?
https://sentry.io/organizations/dimagi/issues/2736738645/
https://sentry.io/organizations/dimagi/issues/2736738409/
They are both happening when I try and download an export from the Case List Explorer.

minhaminha committed:
Fix QA bugs

Broke up export_table into components for passing into task, converted a json incompatible set into a list, fixed a feature breaking typo

minhaminha committed:
Test fix 3

minhaminha committed:
Test fix 4

minhaminha commented:
Apologies to anyone who I accidentally requested for review and for force pushing on a PR! I messed up configuring my branch on a new machine and pushed everything by accident.

gherceg commented on corehq/apps/reports/generic.py:
```diff
@@ -983,25 +949,44 @@ def export_table(self):
+        self.exporting_as_excel = True
+        rows = (self._unformat_row(self, row) for row in self.export_rows)
+        total_row, statistics_rows = [], []
+        if self.total_row:
+            total_row = [self._unformat_row(self, self.total_row)]
+        elif self.statistics_rows:
+            statistics_rows = [self._unformat_row(self, row) for row in self.statistics_rows]
```
Should this be if/elif, or just two separate if statements?

gherceg commented on corehq/apps/reports/generic.py:
```diff
@@ -983,25 +949,44 @@ def export_table(self):
 
         return [[self.export_sheet_name, table]]
 
+    @property
+    def export_table_parts(self):
+        """
+        JSON compatible version of export_table only used for export_all_rows_task calls.
+        export_table will be established in the task with the parts returned from this method.
+        """
```
If the purpose of this method is to return a json compatible exported report, I'd prefer a name like `export_report_as_json`, with the possible refactor of the `export_table` method to be `export_report_as_excel_table`.

Additionally, I don't think you need to mention that this method is only used in the `export_all_rows_task` as that makes this documentation prone to becoming out of date if someone uses this elsewhere. Also a description of what this method returns, similar to what you mentioned in this PR description would be useful.

gherceg commented on corehq/apps/reports/tasks.py:
```diff
@@ -163,33 +167,36 @@ def apps_update_calculated_properties():
+def export_all_rows_task(user_id, report_class, domain, slug, name,
+                         export_format, export_table_parts, recipient_list=None, subject=None):
+    file = io.BytesIO()
+    table = chain(export_table_parts[1], export_table_parts[2])
+    if export_table_parts[3]:
+        table = chain(table, export_table_parts[3])
+    if export_table_parts[4]:
+        table = chain(table, export_table_parts[4])
+    export_table = [[export_table_parts[0], table]]
```
Based on how you are using the indices, I think a more readable/appropriate structure would be to use a dictionary instead of a list. Something like `if 'total_row' in json_report:` is more intuitive than `if json_report[3]:`. Plus, if `json_report` did not contain at least 5 entries for whatever reason, this would raise an exception. 

Also, more of a nit, but similar to my previous naming comment, I'd prefer a different name than `export_table_parts`. My main issue is it sounds like a verb/action, not a variable. I suggested `json_report` because that is my understanding of what this is, but feel free to choose something else.

gherceg reviewed (commented):
Left some comments, mainly concerned about `self.export_table_parts` being a list of lists, rather than a dictionary. I'm also curious if the existing `json_dict` method can be used to pass in a json compatible version of the report.

Also, do you think you could write some tests for the `export_all_rows_task` method? Happy to help out.

minhaminha committed:
Reference the right report object when passing arguments into export_all_rows_task

minhaminha committed:
Use request.POST when request.GET is None

minhaminha commented:
> I'm also curious if the existing `json_dict` method can be used to pass in a json compatible version of the report.

@gherceg I was also curious since I hadn't made any notes about the original arguments passed to `export_all_rows_task` and now I'm kicking myself because I found out that passing both `self.__class__` and `self.__getstate__` work perfectly fine without pickles. I'm not sure what convinced me I needed to go down the path of creating all these new args but I'm going to revert it back to how it was and test on staging to really confirm all I needed to do was take out `serializer='pickle'`, to avoid complicating things (and hopefully fix a bug I haven't been able to figure out).

> Also, do you think you could write some tests for the `export_all_rows_task` method? Happy to help out.

Would it still be worth writing tests after I've reverted it? 

@mjriley Sorry to update you so late on this but I don't think you'd need to make any changes to your celery logging branch. After I've reverted `export_all_rows_task` back to how it was, it'll pass in the request parameters again so `_store_excel_in_blobdb` should work as it did before.

minhaminha committed:
Revert export_all_rows_task to original state

minhaminha committed:
Minor fixes

minhaminha committed:
Merge branch 'master' into ml/remove-pickles-reports

minhaminha committed:
Update export_all_rows_task parameters (testing)

Have export_all_rows_task accept class slugs instead of class type object, delete 'datespan' object before passing into task and reform within it.

minhaminha committed:
Merge branch 'master' into ml/remove-pickles-reports

minhaminha committed:
Re-add Datespan + minor fix

minhaminha committed:
Fix datespan import error

minhaminha committed:
Remove datespan object before passing to send_email_report

minhaminha committed:
Convert to QueryDict + minor fixes

minhaminha committed:
Updated imports

minhaminha commented on corehq/apps/reports/tasks.py:
```diff
@@ -158,9 +160,63 @@ def apps_update_calculated_properties():
+    from corehq.apps.reports.standard.cases.duplicate_cases import DuplicateCasesExplorer
+    from corehq.apps.reports.commtrack.maps import StockStatusMapReport
+    from corehq.apps.reports.commtrack import CurrentStockStatusReport
+    from corehq.apps.reports.commtrack.standard import SimplifiedInventoryReport, InventoryReport
```
will remove eventually

minhaminha committed:
Fix JSONDecodeError + Cleaning up

minhaminha commented:
Another update to this incredibly long PR: 

New QA ticket has been opened [here](https://dimagi-dev.atlassian.net/browse/QA-3872). 

I've updated the PR description to reflect my changes but I'm also going to summarize here too. As mentioned in the last update, I realized the `self.__getstate__` was already JSON compatible (except the `datespan` object) and much less confusing to pass as is than to break it down in a bunch of different pieces, so I reverted `export_all_rows_task` into its former state to do that. `self.__class__` (object type `type`) however was NOT and so I opted instead to import in all report classes this task uses and to identify the correct one using it's `slug` property.

All other tasks in this PR hasn't been changed and still function as they should.

gherceg commented on corehq/apps/reports/generic.py:
```diff
@@ -669,8 +664,12 @@ def export_response(self):
         """
         self.is_rendered_as_export = True
+        report_state = self.__getstate__()
+        report_state['request_params']['startdate'] = str(report_state['request']['datespan'].startdate)
+        report_state['request_params']['enddate'] = str(report_state['request']['datespan'].enddate)
+        del report_state['request']['datespan']
```
2 things. You could use `datespan = report_state['request'].pop('datespan')` to both remove the 'datespan' key/value from the dictionary and reference the variable when setting 'startdate' and 'enddate'. Also it is more explicit to use `.isoformat()` when converting a date to a string.

```suggestion
         datespan = report_state['request'].pop('datespan')
        report_state['request_params']['startdate'] = datespan.startdate.isoformat()
        report_state['request_params']['enddate'] = datespan.enddate.isoformat()
```

gherceg commented on corehq/apps/saved_reports/models.py:
```diff
@@ -818,20 +817,20 @@ def _export_report(self, emails, title):
-            if report_config.report_slug != ApplicationStatusReport.slug:
-                # ApplicationStatusReport doesn't have date filter
+            request_params = json_request(request_data['GET'])
+            if report_config.report_slug != ApplicationStatusReport.slug\
+                    and report_config.report_slug != CaseActivityReport.slug:
```
```suggestion
            if report_config.report_slug not in [ApplicationStatusReport.slug, CaseActivityReport.slug]:
```

gherceg commented on corehq/apps/saved_reports/tasks.py:
```diff
@@ -174,12 +184,16 @@ def send_email_report(self, recipient_emails, domain, report_slug, report_type,
         if getattr(er, 'smtp_code', None) in LARGE_FILE_SIZE_ERROR_CODES or type(er) == ESError:
             # If the email doesn't work because it is too large to fit in the HTML body,
             # send it as an excel attachment.
+            request_params = json_request(request_data['GET'])
+            request_params['startdate'] = str(request_data['datespan'].startdate)
+            request_params['enddate'] = str(request_data['datespan'].enddate)
+            del request_data['datespan']
```
Same here (pop and isoformat)

gherceg commented on corehq/apps/reports/tasks.py:
```diff
@@ -158,9 +160,56 @@ def apps_update_calculated_properties():
-def export_all_rows_task(ReportClass, report_state, recipient_list=None, subject=None):
-    report = object.__new__(ReportClass)
+@task(ignore_result=True)
+def export_all_rows_task(report_class, report_state, recipient_list=None, subject=None):
```
Can you clean this up?

gherceg commented on corehq/apps/reports/tasks.py:
```diff
@@ -158,9 +160,56 @@ def apps_update_calculated_properties():
+    }
+
+    #Reforming datespan object
+    if 'startdate' and 'enddate' in report_state['request_params']:
+        start_date = datetime.strptime(str(report_state['request_params']['startdate']), '%Y-%m-%d %H:%M:%S')
+        end_date = datetime.strptime(str(report_state['request_params']['enddate']), '%Y-%m-%d %H:%M:%S')
```
Since you should be using `.isoformat()` when serializing datetimes, you can use `iso_string_to_datetime` to convert back to datetimes.

gherceg commented on corehq/apps/reports/views.py:
```diff
@@ -521,9 +521,10 @@ def email_report(request, domain, report_slug, dispatcher_class=ProjectReportDis
-
     report_type = dispatcher_class.prefix
-    send_email_report.delay(recipient_emails, domain, report_slug, report_type,
+    request_data['startdate'] = str(request_data['datespan'].startdate)
```
Same here: `request_data['datespan'].startdate.isoformat()`. Personally would use `pop` instead of `del` but up to you

gherceg reviewed (commented):
I think there is still some cleanup that needs to happen on this PR. Did a first pass that remained surface level. Will do a deeper review once the code is cleaned up.

minhaminha committed:
Clean up code + fix test

minhaminha committed:
Merge branch 'master' into ml/remove-pickles-reports

minhaminha committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into ml/remove-pickles-reports

minhaminha committed:
Clean up more code

gherceg commented on corehq/apps/reports/tasks.py:
```diff
@@ -158,9 +160,45 @@ def apps_update_calculated_properties():
-    report = object.__new__(ReportClass)
+@task(ignore_result=True)
+def export_all_rows_task(report_class_name, report_state, recipient_list=None, subject=None):
+    from corehq.apps.reports.standard.monitoring import WorkerActivityReport, DailyFormStatsReport, \
+        CaseActivityReport
+    from corehq.apps.reports.standard.deployments import ApplicationStatusReport
+    from phonelog.reports import DeviceLogDetailsReport
+    from corehq.apps.reports.standard.cases.case_list_explorer import CaseListExplorer
+    from corehq.apps.reports.standard.cases.duplicate_cases import DuplicateCasesExplorer
+    from corehq.apps.reports.commtrack import CurrentStockStatusReport
+    from corehq.apps.smsbillables.interface import SMSBillablesInterface, SMSGatewayFeeCriteriaInterface
+    from corehq.apps.enterprise.interface import EnterpriseSMSBillablesReport
+    from corehq.apps.hqadmin.reports import DeviceLogSoftAssertReport
+
+    reports = {
+        ApplicationStatusReport.slug: ApplicationStatusReport,
+        CaseActivityReport.slug: CaseActivityReport,
+        DailyFormStatsReport.slug: DailyFormStatsReport,
+        DeviceLogDetailsReport.slug: DeviceLogDetailsReport,
+        SMSBillablesInterface.slug: SMSBillablesInterface,
+        SMSGatewayFeeCriteriaInterface.slug: SMSGatewayFeeCriteriaInterface,
+        WorkerActivityReport.slug: WorkerActivityReport,
+        DeviceLogSoftAssertReport.slug: DeviceLogSoftAssertReport,
+        EnterpriseSMSBillablesReport.slug: EnterpriseSMSBillablesReport,
+        CaseListExplorer.slug: CaseListExplorer,
+        DuplicateCasesExplorer.slug: DuplicateCasesExplorer,
+        CurrentStockStatusReport.slug: CurrentStockStatusReport,
+    }
```
If this were the best approach to take, I'd advocate for moving this outside of this method for the sake of readability. Maybe into a const.py file for this module.

However, I don't think you actually need this. The end goal is re-instantiate an object of type `report_class_name`. Rather than pass in the report.slug when calling this task, you can pass in the actual name of the class directly using `report.__class__.__name__` which you can then use below when re-instantiating a report of that type.

gherceg replied:
Ah I may have confused myself. I think it is possible, but not best practice. In that case moving this to const.py would be awesome.

minhaminha replied:
Not sure if this was a concern but I wasn't able to find a way to import all these classes without running into a circular dependency issue so the task will still import every time its called (but now via a function in const.py, so still more readable).

gherceg commented on corehq/apps/saved_reports/models.py:
```diff
@@ -818,20 +817,19 @@ def _export_report(self, emails, title):
-                end_date = datetime.strptime(date_range['enddate'], '%Y-%m-%d')
-                datespan = DateSpan(start_date, end_date)
-                request_data['datespan'] = datespan
+                request_params['startdate'] = str(datetime.strptime(date_range['startdate'], '%Y-%m-%d'))
+                request_params['enddate'] = str(datetime.strptime(date_range['enddate'], '%Y-%m-%d'))
```
Similar to before, calling `<datetime_object>.isoformat()` instead of `str(<datetime_object>)` would be more preferable.

gherceg commented on corehq/apps/accounting/filters.py:
```diff
@@ -237,7 +237,10 @@ def filter_context(self):
     def get_date_str(cls, request, date_type):
-        return request.GET.get('%s_%s' % (cls.slug, date_type))
+        date_str = request.GET.get('%s_%s' % (cls.slug, date_type))
+        if date_str is None:
+            date_str = request.POST.get('%s_%s' % (cls.slug, date_type))
+        return date_str
```
Was this an existing bug that was found, or was there a chance in this PR that resulted in a need for checking POST requests as well?

minhaminha replied:
I can't remember exactly which report it was (probably `CaseActivity` and `ApplicationStatusReport` which don't have date ranges so are handled differently in a few places) but yep, I noticed that for some reason they didn't have a `date_str` in their `GET` but one in their `POST`.

gherceg commented on corehq/apps/reports/generic.py:
```diff
@@ -193,7 +192,7 @@ def __str__(self):
     def __getstate__(self):
         """
-            For pickling the report when passing it to Celery.
+            Returns report parameters to rebuild report in Celery.
```
__getstate__ and __setstate__ are methods specifically used for pickling, so I think now that we are moving away from that (and assuming we don't pickle reports anywhere else), it would be cool to name these methods more appropriately. I also think it is important to note whether in the method name itself or the docs for the method that it returns a json serializable dictionary.

minhaminha replied:
By the method's docs do you mean like the comment on line 195 right now, but more in depth?

gherceg replied:
Yeah just making sure that it is noted the returned dictionary is json serializable
```
def test():
    """
    <method docs>
    """
```

gherceg commented on corehq/apps/reports/generic.py:
```diff
@@ -669,8 +662,12 @@ def export_response(self):
         """
         self.is_rendered_as_export = True
+        report_state = self.__getstate__()
+        datespan = report_state['request'].pop('datespan')
+        report_state['request_params']['startdate'] = datespan.startdate.isoformat()
+        report_state['request_params']['enddate'] = datespan.enddate.isoformat()
```
I think I saw this in one other spot as well, in addition to "reforming the datespan" object after calling `__getstate__`. I think the dictionary returned by `__getstate__` should be usable/json serializable immediately, meaning the `__getstate__` method can handle converting the datespan into a start and end date, and the `__setstate__` method can handle creating a datespan object from a start and end date.

gherceg commented on corehq/apps/saved_reports/tasks.py:
```diff
@@ -174,12 +184,15 @@ def send_email_report(self, recipient_emails, domain, report_slug, report_type,
             # If the email doesn't work because it is too large to fit in the HTML body,
             # send it as an excel attachment.
+            request_params = json_request(request_data['GET'])
+            request_params['startdate'] = request_data['startdate']
+            request_params['enddate'] = request_data['enddate']
```
Similar to my question above, why does this data need to be duplicated?

minhaminha replied:
in `export_all_rows_task` it expects `request_params` to have the date information in order to recreate the datespan object. When refactoring that code to be part of the __setstate__ like you mentioned in another comment, I could feasibly just have it check `request_data` if it can't find it in `request_params`.

minhaminha committed:
Merge branch 'ml/remove-pickles-reports' of github.com:dimagi/commcare-hq into ml/remove-pickles-reports

gherceg commented on corehq/apps/reports/generic.py:
```diff
@@ -670,7 +678,7 @@ def export_response(self):
         self.is_rendered_as_export = True
         if self.exportable_all:
-            export_all_rows_task.delay(self.__class__, self.__getstate__())
+            export_all_rows_task.delay(self.__class__.slug, self.get_json_report_parameters())
```
nit: `self.slug` is sufficient right? I think it works either way.

gherceg reviewed (commented):
I know it is late for this comment and this PR has been sitting for awhile, but can't help but feel like we should take the refactoring opportunity to add some tests. It is a big lift in this area so it might be too much for the scope of this PR, but I worry the longer we put it off the more at risk we are of another report related P1.

I admittedly have a hard time following this area of code which makes me feel ill-equipped to give a meaningful approval. Perhaps @calellowitz or @mjriley could give this another look over.

mjriley reviewed (approved):
The changes look safe, I think my biggest concern is whether we are ever likely to run into issues trying to serialize a QueryDict and not being able to support duplicate keys.

minhaminha committed:
Refactor based on PR feedback

minhaminha commented on corehq/apps/reports/const.py:
```diff
@@ -1,2 +1,32 @@
+
+def get_report_class(slug):
+    # Used by export_all_rows_task to retrieve the correct report class to instantiate
+    from .commtrack import CurrentStockStatusReport
+    from .standard.cases.case_list_explorer import CaseListExplorer
+    from .standard.cases.duplicate_cases import DuplicateCasesExplorer
+    from .standard.deployments import ApplicationStatusReport
+    from .standard.monitoring import WorkerActivityReport, DailyFormStatsReport, CaseActivityReport
+    from corehq.apps.enterprise.interface import EnterpriseSMSBillablesReport
+    from corehq.apps.hqadmin.reports import DeviceLogSoftAssertReport
+    from corehq.apps.smsbillables.interface import SMSBillablesInterface, SMSGatewayFeeCriteriaInterface
+    from phonelog.reports import DeviceLogDetailsReport
```
Yep, I couldn't find a way to have the dictionary stored without a circular dependency error so this is the best I could do. A bit of a hassle to remake it every time but it happens in the task anyway so not too bad

minhaminha commented on corehq/apps/reports/const.py:
```diff
@@ -1,2 +1,32 @@
+        CurrentStockStatusReport.slug: CurrentStockStatusReport,
+    }
+
+    return reports[slug]
```
Added a raise exception line if a slug isn't in the dictionary but at this point with QA having looked over everything, I've accounted for all the reports that would possibly use this. If a new report in gets added to the future the exception will tell them they need to add the slug here.

minhaminha commented on corehq/apps/reports/generic.py:
```diff
@@ -223,20 +227,24 @@ class FakeHttpRequest(object):
 
+        if 'startdate' and 'enddate' in state['request_params']:
+            date_holder = state['request_params']
+        elif 'startdate' and 'enddate' in state['request']:
+            date_holder = state['request']
```
yep! sometimes the start and enddates are in the request and sometimes in the request_params, depending on the report. This could potentially be cleaned up in the future but this is the simpler solution for now.

mjriley replied:
I think I'm confused here. The docstring says that this function will only receive parameters from `get_json_report_parameters`, right? `get_json_report_parameters` does not treat the datespan as optional -- `.pop(property)` raises an exception if the property does not exist. So `report_state['request_params']` will always contain `start_date` and `end_date`, because `get_json_report_parameters` will always set them.

minhaminha replied:
Ahh my bad, that comment isn't strictly true and definitely should've been updated. `export_all_rows_task` which uses the `set_report_parameters` is called from multiple places and the request parameters where the dates aren't in the `request_params` come from when they're called from a task in the saved_reports app, which is what I meant by it could use some cleaning up.

Thinking back on it, I may have pulled the trigger too early on this PR - would it be better to have another PR to clarify that in the comments and/or clean it up so the dates are always in the request_params? Or revert and fix it here?

mjriley replied:
for this specifically, I think a separate PR is fine

minhaminha commented on corehq/apps/reports/views.py:
```diff
@@ -527,9 +527,11 @@ def email_report(request, domain, report_slug, dispatcher_class=ProjectReportDis
+    datespan = request_data.pop('datespan')
+    request_data['startdate'] = datespan.startdate.isoformat()
+    request_data['enddate'] = datespan.enddate.isoformat()
+    send_email_report.delay(list(recipient_emails), domain, report_slug, report_type,
```
nope, sets are not json serializable from what I've seen. I'll add a small comment here

kaapstorm replied:
I see that [`send_email_report`'s docstring](https://github.com/dimagi/commcare-hq/blob/7b5efafd81d7f3d97a60898586c3192c5ef9ff8e/corehq/apps/saved_reports/tasks.py#L113) has `recipient_emails` name as `recipient_list` -- which is a better name, because it hints that the function is expecting a list.

(We should remove that list of parameters from the docstring. All it's doing is explaining their data types, and the only types that are not obvious are `recipient_emails` and `once`. Renaming it `is_once_off` would make the type obvious there too.)

(The reason `recipient_emails` is a set here is to remove duplicates. That wouldn't be necessary if `send_email_report` did its own due diligence. Personally, I would go with
```python
    recipient_list = form.cleaned_data['recipient_emails']
    if form.cleaned_data['send_to_owner']:
        recipient_list.append(request.couch_user.get_email())
```
and then remove duplicates in `send_email_report`. It would be a little simpler, a little more obvious, and there would be no need for comments. (I appreciate this is way out of scope. If I have some time I might just do this a bit later myself.))

minhaminha commented on corehq/apps/reports/generic.py:
```diff
@@ -191,26 +192,29 @@ def __str__(self):
         """
         request = request_as_dict(self.request)
-
-        return dict(
+        report_state = dict(
```
yep it would probably give a JSON error if there was a key with multiple values but after going through all the possible reports we have in cchq, I don't think there's a scenario that would come up anywhere.

minhaminha commented on corehq/apps/reports/tasks.py:
```diff
@@ -158,10 +160,14 @@ def apps_update_calculated_properties():
+@task(ignore_result=True)
+def export_all_rows_task(report_class_name, report_state, recipient_list=None, subject=None):
+    GET_data = QueryDict('', mutable=True)
+    GET_data.update(report_state['request']['GET'])
```
the QueryDict part has been moved to set_report_parameters!

minhaminha commented on corehq/apps/saved_reports/tasks.py:
```diff
@@ -118,6 +120,10 @@ def send_email_report(self, recipient_emails, domain, report_slug, report_type,
+    GET_data = QueryDict('', mutable=True)
+    GET_data.update(request_data['GET'])
+    request_data['GET'] = GET_data
+
```
oops yea, I realized that was kind of redundant so I just shuffled around some of the code so it doesn't do that.

minhaminha commented on corehq/apps/reports/generic.py:
```diff
@@ -223,20 +227,24 @@ class FakeHttpRequest(object):
             datespan = None
             can_access_all_locations = None
 
+        if 'startdate' and 'enddate' in state['request_params']:
```
not impossible! there's reports with just a startdate but they don't need a datespan object if that's case (which is what's being re-made here since datespan objects are not JSON serializable)

minhaminha replied:
thanks for catching that Norman! I said this in another comment but I'm realizing that I may have pulled the trigger on this PR a little too early. Do you suggest I make some sort of amendment PR to address some of these issues or would it be better to revert this PR and continue working on that?

millerdev replied:
Should this PR be reverted? This seems like a bug.

minhaminha replied:
@millerdev I think that would be best since there's other small issues (not bugs like this) that came up after I merged it in too early I'd like to address
